contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+`"ControllerGeneration"` key.
+1. The controller will compute the [hash](#hashing) of the AnyState's `.Data`.
+1. The controller will attach a label to the AnyState so that it is selectable 
+with a low probability of overlap.
+    - ControllerRefs will be used as the authoritative test for ownership.
+    - The specification type Object's `.Selector` should be used where 
+    applicable. 
+    - Alternatively, a Kind unique label may be set to the `.Name` of the 
+    specification type Object.
+1. The controller will add a ControllerRef indicating the specification type 
+Object as the owner of the AnyState in the AnyState's `.OwnerReferences`.
+1. The controller will use the hash from above, along with a user identifiable 
+prefix, to [generate a unique `.Name`](#unique-name-generation) for the AnyState.
+1. The controller will persist the AnyState via the API Server.
+    - Note that, in practice, creation occurs concurrently with 
+    [collision resolution](#collision-resolution).
+    
+### History Reconstruction
+To reconstruct the history of a specification type Object, a controller will do 
+the following. 
+
+1. Select all AnyState Objects labeled as described 
+[above](#version-snapshot-creation).
+1. Filter any AnyStates that do not have a ControllerRef in their 
+`.OwnerReferences` indicating ownership by the Object.
+1. Sort the AnyStates by the value associated with the 
+`"ControllerGeneration"` key.
+1. This produces a strictly ordered set of AnyStates that comprises the ordered 
+revision history of the specification type Object.
+
+### History Maintenance
+Controllers should be configured, either globally or on a per specification type 
+Object basis, to have a `RevisionHistoryLimit`. This field will indicate the 
+number of non-live revisions the controller should maintain in its history
+for each specification type Object. Every time a controller observes a 
+specification type Object it will do the following.
+
+1. The controller will 
+[reconstruct the Object's revision history](#history-reconstruction).
+    - Note that the process of reconstructing the Object's history filters any
+    AnyStates not owned by the Object.
+1. The controller will filter any AnyStates that represent a live version.
+1. If the number of remaining AnyStates is greater than the configured 
+`RevisionHistoryLimit`, the controller will delete them, in order with respect 
+to the value mapped to their `"ControllerGeneration"` keys, until the number 
+of remaining AnyStates is equal to the `RevisionHistoryLimit`.
+
+This ensures that the number of recorded, non-live revisions is less than or 
+equal to the configured `RevisionHistoryLimit`.
+
+### Version Tracking
+Controllers must track the version of the target Object state that corresponds 
+to their generated Objects. This information is necessary to determine which 
+versions are live, and to track which Objects need to be updated during a 
```

Comment:
What do you mean by "and to track which Objects need to be updated during a target state update or rollback"?